src/dash-table/components/EdgeFactory.tsx

```diff
@@ -0,0 +1,216 @@
+            !!filtering,
+            style_cell,
+            style_cell_conditional,
+            style_data,
+            style_data_conditional,
+            style_filter,
+            style_filter_conditional,
+            style_header,
+            style_header_conditional,
+            virtualized.data,
+            virtualized.offset
+        );
+    }
+
+    private memoizedCreateEdges = memoizeOne((
+        columns: VisibleColumns,
+        operations: number,
+        filtering: boolean,
+        style_cell: Style,
+        style_cell_conditional: Cells,
+        style_data: Style,
+        style_data_conditional: DataCells,
+        style_filter: Style,
+        style_filter_conditional: BasicFilters,
+        style_header: Style,
+        style_header_conditional: Headers,
+        data: Data,
+        offset: IViewportOffset
+    ) => {
+        const dataStyles = this.dataStyles(
+            style_cell,
+            style_data,
+            style_cell_conditional,
+            style_data_conditional
+        );
+
+        const filterStyles = this.filterStyles(
+            style_cell,
+            style_filter,
+            style_cell_conditional,
+            style_filter_conditional
+        );
+
+        const headerStyles = this.headerStyles(
+            style_cell,
+            style_header,
+            style_cell_conditional,
+            style_header_conditional
+        );
+
+        let dataEdges = this.getDataEdges(
+            columns,
+            dataStyles,
+            data,
+            offset
+        );
+
+        let dataOpEdges = this.getDataOpEdges(
+            operations,
+            dataStyles,
+            data,
+            offset
+        );
+
+        let filterEdges = this.getFilterEdges(
+            columns,
+            filtering,
+            filterStyles
+        );
+
+        let filterOpEdges = this.getFilterOpEdges(
+            operations,
+            filtering,
+            filterStyles
+        );
+
+        let headerEdges = this.getHeaderEdges(
+            columns,
+            getHeaderRows(columns),
+            headerStyles
+        );
+
+        let headerOpEdges = this.getHeaderOpEdges(
+            operations,
+            getHeaderRows(columns),
+            headerStyles
+        );
+
+        const cutoffWeight = (style_cell ? 1 : 0) + style_cell_conditional.length - 1;
+
+        headerEdges = this.hReconcile(headerEdges, filterEdges || dataEdges, cutoffWeight);
+        headerOpEdges = this.hReconcile(headerOpEdges, filterOpEdges || dataOpEdges, cutoffWeight);
+        filterEdges = this.hReconcile(filterEdges, dataEdges, cutoffWeight);
+        filterOpEdges = this.hReconcile(filterOpEdges, dataOpEdges, cutoffWeight);
+
+        headerOpEdges = this.vReconcile(headerOpEdges, headerEdges, cutoffWeight);
+        filterOpEdges = this.vReconcile(filterOpEdges, filterEdges, cutoffWeight);
+        dataOpEdges = this.vReconcile(dataOpEdges, dataEdges, cutoffWeight);
```

Comment:
As they are cached, these need to be treated as immutable even though they are not. The reconciliation step will clone if required, afterwards the edges are returned as a read-only interface